src/passes/OptimizeInstructions.cpp

```diff
@@ -3994,26 +3994,32 @@ struct OptimizeInstructions
         Binary* add;
         Const* c1;
         Const* c2;
-        if ((matches(curr,
-                     binary(binary(&add, Add, any(), ival(&c1)), ival(&c2))) ||
-             matches(curr,
-                     binary(binary(&add, Add, any(), ival(&c1)), ival(&c2)))) &&
+        if (matches(curr,
+                    binary(binary(&add, Add, any(), ival(&c1)), ival(&c2))) &&
             !canOverflow(add)) {
-          if (c2->value.geU(c1->value).getInteger()) {
-            // This is the first line above, we turn into x > (C2-C1)
-            c2->value = c2->value.sub(c1->value);
-            curr->left = add->left;
-            return curr;
-          }
-          // This is the second line above, we turn into x + (C1-C2) > 0. Other
-          // optimizations can often kick in later. However, we must rule out
-          // the case where C2 is already 0 (as then we would not actually
-          // change anything, and we could infinite loop).
-          auto zero = Literal::makeZero(c2->type);
-          if (c2->value != zero) {
-            c1->value = c1->value.sub(c2->value);
-            c2->value = zero;
-            return curr;
+          // Also check for an overflow when doing C2-C1. We could check this in
+          // a way that considers whether the comparison is signed or unsigned,
+          // but for simplicity for now rule out all overflows in a simple way.
+          auto typeMaxBits = getBitsForType(add->type);
+          auto c1Bits = Bits::getMaxBits(c1, this);
+          auto c2Bits = Bits::getMaxBits(c2, this);
+          if (c1Bits + c2Bits < typeMaxBits) {
```

Comment:
I think it would be pretty simple to perform this check only for signed comparisons to avoid regressing out optimizations of unsigned comparisons.

Reply:
I am actually not 100% sure that is sufficient. This area is rife with with corner cases, and the common usage has small constants anyhow, so I'd prefer to keep this simple. Unless you can see a solution that is easy to prove is always correct?

Reply:
I think the existing code was fairly well-reasoned. It always subtracted a smaller number from a larger number, so it never caused unsigned underflow. We definitely forgot to consider signed underflow, but I don't think there are other edge cases in the unsigned case.

Reply:
Yeah, I guess you're right, I was being overly cautious. Done. I also can't think of another possible overflow issue here (but would not want to bet on it..)

Reply:
friendly ping.